src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -1244,6 +1254,7 @@ internal static unsafe void WalkExternalTrackerObjects()
         {
             bool walkFailed = false;
 
+            // No lock needed to access s_referenceTrackerNativeObjectWrapperCache as this is during GC callback.
```

Comment:
What happens when the thread is suspended for GC in the middle of the s_referenceTrackerNativeObjectWrapperCache update? Can the iteration crash or return inconsistent results?

I think you may want to use a custom-built collection that is specifically hardened for this condition.

Reply:
That is a good point.  From a basic analysis of the current implementation:
- With `List`, there is a risk during removal due to the array copy they do after updating `count`.
- With `HashSet`, that doesn't exist but during `Add`, there is a possibility that we run into one of values still being default due to `Next` being set first which is what is used during iteration.

Do you have any pointers to any existing implementations that I should base the custom implementation on.  There looks to be a very limited version of [ConcurrentSet](https://github.com/dotnet/runtime/blob/9882188fbe16adee3538b65a3926465dfd6052c1/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/ConcurrentSet.cs#L20) today.  But it dosn't support removal and it also looks to be more of a `Dictionary`.

Reply:
I cannot think about existing implementations that are easy to reuse here.

It is not that hard to implement a custom collection with a small number of operations. I think you should try to implement something from scratch.

Reply:
If you would like to validate the approach, you can share the state and structures that you would like to use.

Reply:
Added a custom implementation of a HashSet that should be guarded for enumeration when GC has froze threads.  Let me know if there are any concerns with it.  I did test it with WinUI Control Gallery where we previously saw issues and it seems to be running without hanging / throwing exceptions.